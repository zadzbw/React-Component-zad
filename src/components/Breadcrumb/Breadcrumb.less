@import "../../style/variables";

.zad-bread {
  color: @text-color-secondary;
  font-size: @font-size-base;

  &-item {
    a {
      color: @text-color;
      transition: color @animation-duration-slow;

      &:hover {
        color: tint(@primary-color, 20%);
      }
    }

    &-link {
      > .fa + span {
        margin-left: 0.2857em;
      }
    }

    &:last-child {
      font-weight: bold;
      color: @text-color;
    }

    &-separator {
      margin: 0 8px;
      color: @border-color-base;
    }

    &:last-child &-separator {
      display: none;
    }
  }
}