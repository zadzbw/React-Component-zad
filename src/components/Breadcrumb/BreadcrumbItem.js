/**
 * Created by zad on 16/12/13.
 */
import React, {PropTypes, Component} from 'react';

export default class BreadcrumbItem extends Component {
  constructor(props) {
    super(props);
  }

  static propTypes = {
    separator: PropTypes.string,
    href: PropTypes.string
  };

  static defaultProps = {
    separator: '/'
  };

  render() {
    const {separator, children, ...props} = this.props;
    const link = this.props.href ? (
      <a className="zad-bread-item-link" {...props}>{children}</a>
    ) : (
      <span className="zad-bread-item-link" {...props}>{children}</span>
    );

    return (
      <span className="zad-bread-item">
        {link}
        <span className="zad-bread-item-separator">{separator}</span>
      </span>
    );
  }
}
