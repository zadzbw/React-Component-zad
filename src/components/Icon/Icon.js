import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import isOneOf from '../../utils/isOneOf';

const _type = {
  size: ['lg', '2x', '3x', '4x', '5x'],
  rotate: [90, 180, 270]
};

export default class Icon extends Component {
  static propTypes = {
    name: PropTypes.string.isRequired,
    size: PropTypes.string,
    fixedWidth: PropTypes.bool,
    spin: PropTypes.bool,
    pulse: PropTypes.bool,
    rotate: PropTypes.number
  };

  render() {
    const {name, size, fixedWidth, spin, pulse, rotate, ...props} = this.props;
    const iconClass = classNames('fa', `fa-${name}`, {
      [`fa-${size}`]: size && isOneOf(_type.size, size),
      [`fa-rotate-${rotate}`]: rotate && isOneOf(_type.rotate, rotate),
      'fa-fw': fixedWidth,
      'fa-spin': spin,
      'fa-pulse': pulse
    });
    return (
      <i {...props} className={iconClass}/>
    );
  }
}
