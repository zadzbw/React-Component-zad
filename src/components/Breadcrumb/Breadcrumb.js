/**
 * Created by zad on 16/12/13.
 */
import React, {PropTypes, Component, cloneElement} from 'react';
import classNames from 'classnames';
import BreadcrumbItem from './BreadcrumbItem';

import './Breadcrumb.less';

function getBreadcrumbName(route, params) {
  if (!route.breadcrumbName) {
    return null;
  }

  const paramsKeys = Object.keys(params).join('|');
  // reg: /:(id|name|age)/g
  const reg = new RegExp(`:(${paramsKeys})`, 'g');
  return route.breadcrumbName.replace(reg, (match, p1) => params[p1] || match);
}

function defaultItemRender(route, params, routes, paths) {
  const isLast = routes.indexOf(route) === routes.length - 1;
  const name = getBreadcrumbName(route, params);
  return isLast ? (<span>{name}</span>) : (<a href={`#/${paths.join('/')}`}>{name}</a>);
}

export default class Breadcrumb extends Component {
  constructor(props) {
    super(props);
  }

  static propTypes = {
    routes: PropTypes.array,
    params: PropTypes.object,
    separator: PropTypes.string,
    itemRender: PropTypes.func,
  };

  static defaultProps = {
    routes: [],
    params: {},
    separator: '/',
    itemRender: defaultItemRender
  };

  render() {
    const {routes, params, separator, itemRender, children, className, ...props} = this.props;
    const breadClass = classNames('zad-bread', className);
    let crumbs = '';

    if (routes && routes.length > 0) {
      const paths = [];
      crumbs = routes.map((route) => {
        route.path = route.path || '';
        // 去除开头处的'/'
        let path = route.path.replace(/^\//, '');
        Object.keys(params).forEach((key) => {
          // user/:id => user/123
          path = path.replace(`:${key}`, params[key]);
        });
        if (path) {
          paths.push(path);
        }
        if (route.breadcrumbName) {
          return (
            <BreadcrumbItem separator={separator} key={route.breadcrumbName}>
              {itemRender(route, params, routes, paths)}
            </BreadcrumbItem>
          );
        }
        return null;
      });
    } else if (children) {
      crumbs = React.Children.map(children, (item, index) => {
        if (item.type !== BreadcrumbItem) {
          return null;
        }
        return cloneElement(item, {
          separator,
          key: `bread-crumb-item-${index}`
        });
      });
    }

    return (
      <div {...props} className={breadClass}>
        {crumbs}
      </div>
    );
  }
}
