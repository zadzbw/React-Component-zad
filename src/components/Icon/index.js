/**
 * Created by zad on 16/11/30.
 */
import Icon from './Icon';

export default Icon;