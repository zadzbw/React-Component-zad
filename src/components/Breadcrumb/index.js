/**
 * Created by zad on 16/12/13.
 */
import Breadcrumb  from './Breadcrumb';
import BreadcrumbItem from './BreadcrumbItem';

Breadcrumb.Item = BreadcrumbItem;
export default Breadcrumb;
