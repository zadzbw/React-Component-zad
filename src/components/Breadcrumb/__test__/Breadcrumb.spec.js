/**
 * Created by zad on 16/12/13.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Breadcrumb from '../../Breadcrumb';
import Icon from '../../Icon';

const BreadcrumbItem = Breadcrumb.Item;

describe('Breadcrumb test', () => {
  it('normal', () => {
    const wrapper = shallow(
      <Breadcrumb>
        <BreadcrumbItem>Home</BreadcrumbItem>
        <BreadcrumbItem><a href="#/post">Post</a></BreadcrumbItem>
        <BreadcrumbItem href={'#/app'}><Icon name={'shield'}/><span>Application</span></BreadcrumbItem>
        <div>test</div>
      </Breadcrumb>
    );

    expect(wrapper.hasClass('zad-bread')).toBeTruthy();
    expect(wrapper.children()).toHaveLength(3);

    expect(wrapper.find(BreadcrumbItem).at(0).shallow().text()).toEqual('Home/');
    expect(wrapper.find(BreadcrumbItem).at(0).shallow().find('.zad-bread-item-link').type()).toEqual('span');

    expect(wrapper.find(BreadcrumbItem).at(1).shallow().find('.zad-bread-item-link').type()).toEqual('span');
    expect(wrapper.find(BreadcrumbItem).at(1).shallow().find('.zad-bread-item-link').children().at(0).type()).toEqual('a');
    expect(wrapper.find(BreadcrumbItem).at(1).shallow().find('.zad-bread-item-link').children().get(0).props.href).toEqual('#/post');

    expect(wrapper.find(BreadcrumbItem).at(2).shallow().find('.zad-bread-item-link').type()).toEqual('a');
    expect(wrapper.find(BreadcrumbItem).at(2).shallow().find('a').at(0).hasClass('zad-bread-item-link')).toBeTruthy();
    expect(wrapper.find(BreadcrumbItem).at(2).shallow().find('span').at(0).hasClass('zad-bread-item-link')).toBeFalsy();

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('separator', () => {
    const wrapper = shallow(
      <Breadcrumb separator={'>'}>
        <BreadcrumbItem>Home</BreadcrumbItem>
        <BreadcrumbItem>Application</BreadcrumbItem>
        <BreadcrumbItem>About</BreadcrumbItem>
      </Breadcrumb>
    );

    expect(wrapper.find(BreadcrumbItem).at(0).shallow().text()).toEqual('Home>');
    expect(wrapper.find(BreadcrumbItem).at(1).shallow().text()).toEqual('Application>');
    expect(wrapper.find(BreadcrumbItem).at(2).shallow().text()).toEqual('About>');

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('react-router mock', () => {
    const mock = {
      bread: {
        // react-router mock
        params: {userId: 1, appId: 2, postId: 3},
        routes: [
          {path: '/', breadcrumbName: 'Home'},
          {path: 'user/:userId', breadcrumbName: 'User:userId'},
          {path: 'application/:appId', breadcrumbName: 'Application:appId'},
          {path: 'post/:postId', breadcrumbName: 'Post:postId'}
        ]
      }
    };
    const wrapper = shallow(<Breadcrumb params={mock.bread.params} routes={mock.bread.routes}/>);

    expect(wrapper.find(BreadcrumbItem).at(0).shallow().text()).toEqual('Home/');
    expect(wrapper.find(BreadcrumbItem).at(0).shallow().find('.zad-bread-item-link').children().get(0).props.href).toEqual('#/');

    expect(wrapper.find(BreadcrumbItem).at(1).shallow().text()).toEqual('User1/');
    expect(wrapper.find(BreadcrumbItem).at(1).shallow().find('.zad-bread-item-link').children().get(0).props.href).toEqual('#/user/1');

    expect(wrapper.find(BreadcrumbItem).at(2).shallow().text()).toEqual('Application2/');
    expect(wrapper.find(BreadcrumbItem).at(2).shallow().find('.zad-bread-item-link').children().get(0).props.href).toEqual('#/user/1/application/2');

    expect(wrapper.find(BreadcrumbItem).at(3).shallow().text()).toEqual('Post3/');
    expect(wrapper.find(BreadcrumbItem).at(3).shallow().find('.zad-bread-item-link').children().at(0).type()).toEqual('span');
    expect(wrapper.find(BreadcrumbItem).at(3).shallow().find('.zad-bread-item-link').children().text()).toEqual('Post3');

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
