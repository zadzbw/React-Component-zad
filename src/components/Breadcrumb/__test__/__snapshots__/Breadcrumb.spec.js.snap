exports[`Breadcrumb test normal 1`] = `
<div
  className="zad-bread">
  <BreadcrumbItem
    separator="/">
    Home
  </BreadcrumbItem>
  <BreadcrumbItem
    separator="/">
    <a
      href="#/post">
      Post
    </a>
  </BreadcrumbItem>
  <BreadcrumbItem
    href="#/app"
    separator="/">
    <Icon
      name="shield" />
    <span>
      Application
    </span>
  </BreadcrumbItem>
</div>
`;

exports[`Breadcrumb test react-router mock 1`] = `
<div
  className="zad-bread">
  <BreadcrumbItem
    separator="/">
    <a
      href="#/">
      Home
    </a>
  </BreadcrumbItem>
  <BreadcrumbItem
    separator="/">
    <a
      href="#/user/1">
      User1
    </a>
  </BreadcrumbItem>
  <BreadcrumbItem
    separator="/">
    <a
      href="#/user/1/application/2">
      Application2
    </a>
  </BreadcrumbItem>
  <BreadcrumbItem
    separator="/">
    <span>
      Post3
    </span>
  </BreadcrumbItem>
</div>
`;

exports[`Breadcrumb test separator 1`] = `
<div
  className="zad-bread">
  <BreadcrumbItem
    separator=">">
    Home
  </BreadcrumbItem>
  <BreadcrumbItem
    separator=">">
    Application
  </BreadcrumbItem>
  <BreadcrumbItem
    separator=">">
    About
  </BreadcrumbItem>
</div>
`;
